WifiART/src/userinterface/LcrModel.java

```diff
@@ -60,10 +81,73 @@ public void setCallback(LcrController fc) {
         this.fc = fc;
     }
 
-
-    // TODO(dmevans) Implement the encoding in toHexBuffer().
     @Override
     public String toHexBuffer() {
-        return null;
+        // Initialize capacity of buffer to the minimum buffer size.
+        List<Byte> byteBuffer = new ArrayList<>(ADDRESS_ELEMENT_LIST_INDEX);
+        byteBuffer.add(SUBELEMENT_ID_INDEX, SUBELEMENT_ID); // Byte 0
+        byteBuffer.add(LENGTH_INDEX, (byte) 0); // Byte 1, placeholder for length field
+
+        // Encode the country,
+        String country = state.getCountry();
+        try {
+            String countryCode = CountryCodes.COUNTRY_NAMES_TO_CODES_MAP.get(country);
+            fillStringBytes(byteBuffer, countryCode); // Bytes 2-3
+        } catch (NullPointerException exception) {
+            throw new NullPointerException(COUNTRY_NOT_CHOSEN);
+        }
+
+        // Encode the address elements.
+        List<AddressElement> addressElementsList = state.getAddressElementsList();
+        // Comparator is null because Address Element implements Comparable.
+        addressElementsList.sort(null);
+        String currentLanguage = "";
+        for (AddressElement addressElement : addressElementsList) {
+            String languageCode = addressElement.getLanguageCode();
+            if (!(languageCode.equals(currentLanguage))) {
+                byteBuffer.add(LANGUAGE_CA_TYPE);
+                byteBuffer.add(LANGUAGE_CODE_LENGTH);
+                fillStringBytes(byteBuffer, languageCode);
+                currentLanguage = languageCode;
+            }
+            byteBuffer.add(addressElement.getCivicAddressType());
+            String addressElementName = addressElement.getName();
+            int lengthIndex = byteBuffer.size();
+            byteBuffer.add((byte) 0); // Placeholder for the length
+            int addressElementLength = fillStringBytes(byteBuffer, addressElementName);
+            if (addressElementLength > MAX_BYTE_VALUE) {
+                throw new IllegalArgumentException(ADDRESS_IS_TOO_LONG);
```

Comment:
This should be IllegalStateException I think. IllegalArgumentException is used when the method has been passed with an illegal argument, which isn't the case.

Reply:
I'm not sure IllegalStateException would be applicable here though, because there is nothing about the state that would determine that the method is being called incorrectly. I think there is a sense in which IllegalArgumentException makes sense in that the user input was an illegal argument. Maybe IndexOutOfBounds is more appropriate because this error occurs when the address element is too long to be encoded in the maximum-length buffer.